Add rocket chat import tool

garg3133 opened the pull request:
This PR adds the support to import the following data from Rocket Chat:
* All users
* All public/private channels
* All teams and their public/private channels
* All discussion rooms as topics in their parent channel
* All the messages in all the channels
* All private conversations
* Reactions on messages (except for custom emojis)
* Mentions in messages (except `@all`, `@here` mentions)

Features remaining to be added:
* User avatars
* Attachments
* Default streams
* Custom emojis
* `@all`, `@here` mentions
* Starred messages
* Threads as separate topics
* Import messages longer than 10,000 characters.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,675 @@
+        user_dict = user_id_to_user_map[rc_user_id]
+        user_dict["is_mirror_dummy"] = False
+
+        if user_dict["type"] != "user":
```
Can you add a comment documenting the other types that might exist in Rocket.Chat?

garg3133 replied:
Added the comment.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,675 @@
+        content = content.replace("@all", "@**all**")
+        # We don't have an equivalent for Mattermost's @here mention which mentions all users
+        # active in the channel.
+        content = content.replace("@here", "@**all**")
```
This comment is kinda confusing; we should just support exactly the set of wildcard mentions that exist in Rocket.Chat, without reference to what Mattermost happens to do.

garg3133 replied:
I actually copied the comment from the Mattermost tool, which is " We don't have an equivalent for Mattermost's @here mention which mentions all users online in the channel.". I changed the `online` to `active`, as RocketChat tags all the active users in a channel with `@here`, but forgot to change the name.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,675 @@
+            mention_user_ids=mention_user_ids,
+        )
+
+        if len(content) > 10000:  # nocoverage
```
It's probably better to truncate too-long messages; we could also consider spitting them.  Did you see this case in practice, or is this copied from Mattermost?  If the latter, probably best to leave this for a follow-up issue.

garg3133 replied:
I copied this from the Mattermost import tool, but I checked for RocketChat and it also supports sending messages longer than 10,000 characters.

I've added this in the description of this PR, to be solved in a follow-up issue.

timabbott replied:
Sounds good.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,679 @@
+import logging
```
Is there top-level documentation on the Rocket.Chat data export format that we could link to from a comment located here?

timabbott replied:
(Might be worth adding something similar for the other data import tools, if we haven't already).

garg3133 replied:
There isn't any such documentation for Rocket.Chat (they don't even have any mechanism to export data, we use the mongo dump in our import tool). But it would be worth adding this in other import tools.

iammattmartin commented:
Should this already work for the initial list of data?

I'm seeing `if message["rid"] in direct_channels_list:` on a first run.

garg3133 commented:
@iammattmartin Thanks for reporting this. Would it be possible for you to add a `print(message)` statement before this line, i.e., line 432, and share the message object for which the error is occurring? I'm not sure what kind of message this is which is not having a room_id.

iammattmartin commented:
Sure.

```
{'_id': 'TNFWJeb6ivXkkzFcL', 't': 'room_changed_privacy', 'rid': 'cGbD4N6B4QAZYSf2T', 'ts': datetime.datetime(2019, 2, 22, 1, 55, 58, 787000), 'msg': 'Private Group', 'u': {'_id': 'ztSfEeAh5uaFYLLWq', 'username': 'xxxx'}, 'groupable': False, '_updatedAt': datetime.datetime(2019, 11, 6, 8, 59, 31, 295000)}
```

garg3133 commented:
@iammattmartin But this does have a `rid` key. The message object causing error shouldn't have that key and it must be printed just before the error.

iammattmartin commented:
Ah sorry, here is the full error.

```
Converting Data ...
Traceback (most recent call last):
  File "./manage.py", line 157, in <module>
    execute_from_command_line(sys.argv)
  File "./manage.py", line 122, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip/zerver/management/commands/convert_rocketchat_data.py", line 45, in handle
    do_convert_data(rocketchat_data_dir=data_dir, output_dir=output_dir)
  File "/srv/zulip/zerver/data_import/rocketchat.py", line 621, in do_convert_data
    private_messages=private_messages,
  File "/srv/zulip/zerver/data_import/rocketchat.py", line 432, in separate_channel_and_private_messages
    if message["rid"] in direct_channels_list:
KeyError: 'rid'
```

garg3133 commented:
Is this what got logged after including the `print(message)` statement above line 432? I just want the last message that got printed just before the error trace, to see what kind of message that is that does not have a `rid`, i.e., it doesn't belong to any room.

iammattmartin commented:
The output was quite large, (many hundreds of lines) but last line above the traceback was:

{'_id': 'p4v37myxc6yLZ8AHh', 't': 'livechat_navigation_history', 'ts': datetime.datetime(2019, 11, 6, 0, 38, 42, 796000), 'msg': ' - applewebdata://9124F033-BFEF-43C5-9215-DA369E4DA22D', 'u': {'_id': 'rocket.cat', 'username': 'cat'}, 'groupable': False, 'unread': True, 'navigation': {'page': {'change': 'url', 'title': '', 'location': {'href': 'applewebdata://9124F033-BFEF-43C5-9215-DA369E4DA22D'}}, 'token': 'ebxuypgh0updo6klkobzhp'}, 'expireAt': 1575592722794.0, '_hidden': True, '_updatedAt': datetime.datetime(2019, 11, 6, 0, 38, 42, 796000)}
```

We never actively used this and deleted it but seems that some data is never actually removed from Rocket.Chat

iammattmartin commented:
On testing a bit more removing the livechat entries, we now get:

```
Traceback (most recent call last):
  File "./manage.py", line 157, in <module>
    execute_from_command_line(sys.argv)
  File "./manage.py", line 122, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip/zerver/management/commands/convert_rocketchat_data.py", line 45, in handle
    do_convert_data(rocketchat_data_dir=data_dir, output_dir=output_dir)
  File "/srv/zulip/zerver/data_import/rocketchat.py", line 661, in do_convert_data
    output_dir=output_dir,
  File "/srv/zulip/zerver/data_import/rocketchat.py", line 390, in process_messages
    raw_messages.append(message_to_dict(message))
  File "/srv/zulip/zerver/data_import/rocketchat.py", line 337, in message_to_dict
    reactions = list_reactions(message["reactions"])
  File "/srv/zulip/zerver/data_import/rocketchat.py", line 320, in list_reactions
    for react_code in reactions:
TypeError: 'NoneType' object is not iterable
```

Printing the last error gives:

```
{'_id': '4MYyEiVTk3vLBsJLcW', 'rid': 'TPX9byRJwy4Wmsh5tztSfEeAh5uaFYLLWq', 'msg': 'Message text here', 'ts': datetime.datetime(2018, 9, 25, 21, 3, 18, 704000), 'alias': 'Person Name', 'u': {'_id': 'TPX9byRJwy4Wmsh5t', 'username': 'user', 'name': 'Person Name'}, '_updatedAt': datetime.datetime(2018, 9, 25, 21, 3, 30, 211000), 'editedBy': None, 'editedAt': None, 'emoji': None, 'avatar': None, 'customFields': None, 'attachments': None, 'mentions': [], 'channels': [], 'reactions': {':joy:': {'usernames': ['username2']}}}
```

Looks like I won't be able to remove those as easy.

garg3133 commented:
Updated code to ignore the messages which do not belong to any channel (these messages generally represent server-wide events) and use conditionals of form `if dict_name.get("key_name")` instead of `if "key_name" in dict_name`, as the latter passes if there's an item in the dictionary of the form `dict_name["key_name"] = None`, which we don't want.

@iammattmartin Can you try running this again?

iammattmartin commented:
Looking good.

```
2021-07-12 15:12:56.878 INFO [] Start making tarball
2021-07-12 15:12:57.030 INFO [] Done making tarball
```

Imported and no errors, but will let you know if we see any other oddities.

iammattmartin commented:
Final note, all data seems to look good. Can't comment on the bits unsupported but we didn't use attachments, threads or starred messages.

Many thanks for this!

garg3133 commented:
Added test coverage for this PR.

The CI failure is unrelated to this PR.
```
ERROR: zerver/models.py no longer has complete backend test coverage
  Lines missing coverage: [869]
```

timabbott commented on zerver/management/commands/convert_rocketchat_data.py:
```diff
@@ -0,0 +1,45 @@
+
+        if os.listdir(output_dir):
+            raise CommandError("Output directory should be empty!")
+        output_dir = os.path.realpath(output_dir)
```
A lot of this code is duplicated between our conversion tools; what do you think about doing a small prep PR to migrate the existing import tools to a cleaner interface?  We can start by using `ZulipBaseCommand` from `zerver/lib/management.py`, and then adding a `self.add_data_export_output_args()` similar to `add_realm_args`, and then call a `selfoutput_dir = self.get_output_dir()` that together replace most of the code from here on up.

(The other bits seem harder to avoid duplicating, and are relatively short in any case).

timabbott commented on zerver/tests/test_rocketchat_importer.py:
```diff
@@ -0,0 +1,672 @@
+        self.assertEqual(zerver_stream[1]["invite_only"], True)
+        self.assertEqual(zerver_stream[1]["description"], "")
+        self.assertEqual(zerver_stream[1]["rendered_description"], "")
+        self.assertEqual(zerver_stream[1]["realm"], 3)
```
Let's use `realm_id` rather than `3` for these checks; that's a lot more readable.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+def make_realm(realm_id: int, realm_subdomain: str, domain_name: str) -> ZerverFieldsT:
+    NOW = float(timezone_now().timestamp())
+
+    zerver_realm = build_zerver_realm(realm_id, realm_subdomain, NOW, "RocketChat")
```
I think we should spell this last parameter `Rocket.Chat`, since that's how they spell their name.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+
+
+def make_realm(realm_id: int, realm_subdomain: str, domain_name: str) -> ZerverFieldsT:
+    NOW = float(timezone_now().timestamp())
```
If the Rocket.Chat export has a realm creation time, it's possible we should record that instead of the current time.  

(This may be an issue across our data import tools)

garg3133 replied:
Mattermost does not provide any timestamps (realm creation and stream creation) and Slack only provides the stream creation timestamp, which is been correctly used while creating Zulip streams. So, this is a problem with Zulip only.

Will fix it.

garg3133 replied:
Update code to use the actual realm creation and stream creation time.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+        # "bot": A special user types for bots.
+        # "unknown": This usually represents a livechat guest.
+        if user_dict["type"] != "user":
+            user_dict["is_mirror_dummy"] = True
```
I think we should only set `is_mirror_dummy` for `"unknown"`.  

We probably want to import bots with `is_bot=True, is_active=False`, and ideally `bot_owner` set.

garg3133 replied:
Done.

In Rocket.Chat, there is nothing like `bot_owner` and all the admins have the privilege to configure all the bots. But, as we can set only one user as the `bot_owner`, while looping through all the users (when creating their user profiles), I separated out all the admins (now realm owners) in `realm_owners` list and set the first realm owner as the owner of all the bots.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+) -> None:
+    for rc_user_id in user_id_to_user_map:
+        user_dict = user_id_to_user_map[rc_user_id]
+        user_dict["is_mirror_dummy"] = False
```
I'm not sure why you need this variable as a parameter on `user_dict`, rather than just a local variable, given that you convert it into one below.

garg3133 replied:
Ohh, I must have missed this.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+            else:
+                stream_desc = "[Team {} channel]. {}".format(
+                    team_id_to_team_map[channel_dict["teamId"]]["name"], stream_desc
+                )
```
What's the thinking for the `[TEAM] {stream_name}` formatting of the team channel name?  I think it'd be reasonable to not change stream names and just do something with the stream descriptions.  But maybe I'm missing something about how "teams" work in Rocket.Chat and the stream names aren't themselves globally unique?

timabbott replied:
If we need to combine the channel and team names for uniqueness, I think what I'd want is a format for stream names like `#{channel_name} ({team})`

garg3133 replied:
Actually, teams have a separate category in Rocket.Chat, where all the main channels of teams are grouped together. (All teams have one main channel that appears in that category and then several child channels that appears in the normal `Public` and `Private` category.)

![Screenshot from 2021-07-20 03-13-53](https://user-images.githubusercontent.com/39924567/126231428-5782dc2a-9e25-4f06-a02f-1cbb23f0c8ce.png)


So, in order to give the users migrating from Rocket.Chat a similar experience, I added `[TEAM]` before the stream name of the main team channels, which will group all the teams together at the top of the streams list, just like in Rocket.Chat.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+        invite_only = channel_dict["t"] == "p"
+
+        stream_name = channel_dict["name"]
+        stream_desc = channel_dict.get("description", "")
```
Do they have an equivalent of our `stream_post_policy` that we could import?  https://zulip.com/help/stream-sending-policy.

garg3133 replied:
I looked into it and yes, on Rocket.Chat we can mark channels as read-only, where only admins of the workplace and owners and moderators of the channel can post.

While importing such channels, I set the stream policy as -- Only admins and moderators can post.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+            name=stream_name,
+            description=stream_desc,
+            stream_id=stream_id,
+            deactivated=False,
```
I think we might need `deactivated=True` for streams with no subscribers, since Zulip auto-deactivates streams in that situation.

garg3133 replied:
Done.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+            emoji_code = name_to_codepoint[emoji_name]
+            reaction_type = Reaction.UNICODE_EMOJI
+        else:  # nocoverage
+            continue
```
I take it we haven't implemented custom emoji import yet?

garg3133 replied:
Yes, removed all the comments related to custom emojis.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+                zulip_member_id = user_id_mapper.get(rc_member_ids[0])
+                message_dict["recipient_id"] = user_id_to_recipient_id[zulip_member_id]
+            # PMs don't have topics, but topic_name field is required in `build_message`.
+            message_dict["topic_name"] = "Imported from rocketchat"
```
If this is a PM, `topic_name` should be the empty string.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+        else:
+            stream_id = stream_id_mapper.get(message["rid"])
+            message_dict["recipient_id"] = stream_id_to_recipient_id[stream_id]
+            message_dict["topic_name"] = "Main channel content (imported from rocketchat)"
```
And then this can be just "Imported from Rocket.Chat" to be consistent with our other import tools.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+            parent_channel_id = dsc_channel["prid"]
+            stream_id = stream_id_mapper.get(parent_channel_id)
+            message_dict["recipient_id"] = stream_id_to_recipient_id[stream_id]
+            message_dict["topic_name"] = "(Discussion) {}".format(dsc_channel["fname"])
```
Maybe "{fname} (from Rocket.Chat)` would be a good pattern here?  I'm not sure what a `fname` looks like.

garg3133 replied:
I think we should keep it as `{fname} (Discussion from Rocket.Chat)`. This way, it won't take a good portion of the topic name at the beginning and will also inform the users that this topic is actually imported from Rocket.Chat Discussion.

You can refer to the screenshot attached above to get an idea of how the `fname` looks like (these are just like the stream names).

timabbott replied:
maybe `{fname} (Imported from Rocket.Chat)`?  I think that's a bit cleaner.

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+                continue
+            user_id = user_id_mapper.get(mention_id)
+            mention_user_ids.add(user_id)
+        message_dict["mention_user_ids"] = mention_user_ids
```
What does setting `mentions_user_ids` here do?  I think the markdown rendering process during data import will handle individual user mentions, if we've updated the content, and this code may not do anything.

garg3133 replied:
> I think the markdown rendering process during data import will handle individual user mentions

No, it does not. 

So, while creating the user_message, we just add `8` to the `flag_mask` (to state that the user is mentioned in the message). This highlights the message for the users mentioned in the message but does not highlight the part where the users are mentioned in the message, plus, clicking over the user mention in the message does not open any popover.

![Screenshot from 2021-07-24 03-42-50](https://user-images.githubusercontent.com/39924567/126846333-c57bbaed-a6c0-40c4-8954-50491917fa29.png)

timabbott commented on zerver/data_import/rocketchat.py:
```diff
@@ -0,0 +1,673 @@
+    for message in messages:
+        if message.get("t") is not None:
+            # Message contains user_joined, added_user, discussion_created,
+            # etc. feeds.
```
Be explicit: "Messages with a type are system notifications like user_joined that we don't include."

timabbott commented:
@garg3133 this looks great!  In addition to resolving the inline notes above, we should add a /help/import-from-rocketchat article similar to the existing ones for Slack/Mattermost; that's worth doing in this PR since it's a place to document caveats like the lack of support for custom emoji and uploaded files.

(BTW, I think rather than commenting out the custom emoji code, we'd be better off just having it be an empty array at the top-level function, so the plumbing doesn't need to be cluttered with commented blocks that likely work).

timabbott commented:
@garg3133 can you rebase this, so that it drops the already-merged commits from the PR?

timabbott commented:
Also, it'd be great to have a long comment with the overall status; I think we're close to merging but am not sure where my attention would be best spent on further work.

@hackerkid I'd like your thoughts from another round of review as well.

garg3133 commented:
At this point, this PR majorly misses the documentation part, which I plan to add this weekend. I have updated the code resolving all other review comments, except for the `stream_post_policy` which I haven't looked into yet.

All the other remaining review comments (the comment at the top, first comment of second review and the `mention_user_ids` thing) involve changes that affect all the import tools and are better done as separate PRs.

garg3133 commented:
I've added the `stream_post_policy` as well for the read-only channels in Rocket.Chat. While looking for it in the Rocket.Chat docs, I also find out that Rocket.Chat also supports multi-user direct messaging and archiving a stream.

I've added the support for huddles in the To-Do list above (in PR description) but I'm not sure about the equivalent for archive feature in Zulip.

According to Rocket.Chat docs:
> **Archived:** Tells if a channel is archived and toggles the archived status of a channel. Nobody can post messages in an archived channel, and channel search will not find that channel.

garg3133 commented:
I've added documentation for Rocket.Chat as well, please take a look.

garg3133 commented:
Added support for importing huddles as well, since it is an important feature and should be available on the first merge.

hackerkid reviewed (commented):
Hey, thanks for working on this. I tested this by making an import using the fixture data present in the tests. SOme minor comments
- It looks like links to the stream are broke.
  ![Screenshot from 2021-07-28 17-55-30](https://user-images.githubusercontent.com/7190633/127336907-4b40e5f5-84b5-42f7-a46e-d2e92ae16d94.png)
- I think the current convention of how we handle the channels that belong to teams can be improved. For context, you can create teams in mattermost within an organization.  Each and every team has a main channel and can have additional channels as well. The channels can have users outside the team. And a user can be part of multiple teams.
Currently we mention whether a channel is part of the team in description. Unless it's the main channel.
![Screenshot from 2021-07-28 14-07-16](https://user-images.githubusercontent.com/7190633/127336986-6c761353-eadc-430f-9cff-d434e7631287.png)
   I think there are two main issues because of this.
	- We loose the original description of the stream.
	-  It's not at all obvious from the main stream menu on the leftside of the app which stream belongs to which team. Could probably even a bigger issue if two or more teams have channels with the same name since [users can be part of multiple teams](https://docs.rocket.chat/guides/user-guides/teams/teams-members).	
![Screenshot from 2021-07-28 14-08-57](https://user-images.githubusercontent.com/7190633/127337198-c7a0c1cf-d3f0-46f5-87af-f677ee0fc80c.png)
	   For example multiple teams can have a channel  called docs
![image (2)](https://user-images.githubusercontent.com/7190633/127337254-bc31dab9-af5e-4c96-b3ea-6dcee1b36782.png)
	- I think the solution might to have a naming convention like this
		- For the main channel let's call it - `{team_name} - main`
		- For other channels lets call it - `{team_name} - {channel_name}`
		- And we keep the original description
- It looks like we set a random realm owner as the owner of all the bots since we don't have that information? I think it's probably fine for now. Maybe let's make it obvious that the bot is imported from RC by mentioning it in the name or email?
- What does the following mean? Does it means the user used to be part of team-harry-potter and thp-channel-2 and no longer does? Since the topic is mentioned in the room and not the actual stream itself. At the moment we treat it lilke the user is not subscribed to the stream right? Was that the case when you created the import as well?
```python
"__rooms": [
	"GENERAL",
	"random",
	"team-harry-potter > t-h-p discussion",
	"thp-channel-2 > thp-channel-2 discussions",
	"random > random discussion"
],
```

garg3133 commented:
> It looks like links to the stream are broke.

Yes, as the markdown of the messages is not processed in the backend during the import, mentions and links to other streams are broken. For mentions, we explicitly mark for each user (if this user is mentioned in this message or not) during the run of this tool only, so clicking over the mentions do not open their profile popovers.

> Currently we mention whether a channel is part of the team in description. Unless it's the main channel.
> I think there are two main issues because of this.
> - We loose the original description of the stream.

No, we don't lose the original description, it is just prepending the `[Team {team_name} channel].` to the actual description of the team channel. It is not appearing here because those channels actually don't have any channels.

> - It's not at all obvious from the main stream menu on the leftside of the app which stream belongs to which team.

This is actually true for Rocket.Chat too. In Rocket.Chat, all the normal public-private channels are mixed up with teams' public-private channels, so you can't tell just by looking at the channels in the left sidebar if the channel is a normal channel or it belongs to a team, and if it does belong to a team, to which team? You can only tell this by first opening the channel and looking at the channel header, where the team name is mentioned.

![rocketchat](https://user-images.githubusercontent.com/39924567/126231428-5782dc2a-9e25-4f06-a02f-1cbb23f0c8ce.png)

![Screenshot from 2021-07-28 20-34-07](https://user-images.githubusercontent.com/39924567/127346425-2f3a3d0d-16b0-43e3-8659-3a79634eff5d.png)

> Could probably even a bigger issue if two or more teams have channels with the same name since users can be part of multiple teams.

Yes, users can be a part of multiple teams but still, no two teams can have channels with the same name in Rocket.Chat. All the channel names appearing in the left sidebar (whether it is a team's main channel, team's public/private channel or a normal channel) have unique names.


> - I think the solution might to have a naming convention like this
      * For the main channel let's call it - {team_name} - main
      * For other channels lets call it - {team_name} - {channel_name}
      * And we keep the original description

Yes, we can do this but it will just make the streams look different to how they looked in Rocket.Chat (users might be used to how the channels were arranged in Rocket.Chat and such a change in channel arrangement might get them confused?):
* Teams' main channels are no longer together and at the top of stream list (this was the main reason of pretending `[TEAM]` to the teams' main channel names.
* Teams' channels names will change in the left sidebar and `{team_name}` might happen to occupy a large space in the left sidebar and then the actual channel name mighe not be visible properly.

garg3133 commented:
> What does the following mean? Does it means the user used to be part of team-harry-potter and thp-channel-2 and no longer does? Since the topic is mentioned in the room and not the actual stream itself. At the moment we treat it lilke the user is not subscribed to the stream right? Was that the case when you created the import as well?
```
"__rooms": [
	"GENERAL",
	"random",
	"team-harry-potter > t-h-p discussion",
	"thp-channel-2 > thp-channel-2 discussions",
	"random > random discussion"
],
```

This means that the user is subscribed to the discussions `t-h-p discussion` and `thp-channel-2 discussions` in Rocket.Chat, but is not subscribed to the channels to which these discussions belong. For the `random` channel, the user is subscribed to the channel as well as one of its discussions.

For such cases, while importing the discussions as a topic in their parent stream, we do not subscribe such users to the streams. After importing the organization to Zulip, admins would have to decide themselves whether to subscribe such people to the whole stream, leave those people from the discussion or just move the topic to a new stream to which the users who are left out can be unsubscribed.

I should have mentioned all this in the Ceveats section of the documentation, my bad!

hackerkid commented:
>  In Rocket.Chat, all the normal public-private channels are mixed up with teams' public-private channels, so you can't tell just by looking at the channels in the left sidebar if the channel is a normal channel or it belongs to a team, and if it does belong to a team, to which team? You can only tell this by first opening the channel and looking at the channel header, where the team name is mentioned.

hmm. I would have imagined they must have a better UI than that. If they don't explicitly show the team name either then I guess it's probably fine for us to not show that in the stream name as well.

> Yes, users can be a part of multiple teams but still, no two teams can have channels with the same name in Rocket.Chat. 

Ahh. I got confused by their documentation which showed "docs" as one of the example channels they have shown in the their documentation demonstrating the "team" feature. Kind of a misleading example since, docs is the exactly the kind of channel which multiple teams can have.

https://docs.rocket.chat/guides/user-guides/teams/teams-channels/create-new-channel-for-a-team

> This means that the user is subscribed to the discussions t-h-p discussion and thp-channel-2 discussions in Rocket.Chat, but is not subscribed to the channels to which these discussions belong. For the random channel, the user is subscribed to the channel as well as one of its discussions.

Cool. Thanks for the clarification. I wonder whether we should subscribe the users to the channels in that case? Better they unsubscribe themselves later than they missing some probably important messages right?

garg3133 commented:
> Better they unsubscribe themselves later than they missing some probably important messages right?

But if the channel is private, wouldn't that lead them to access the messages they are not supposed to see? Isn't there a way to somehow just notify the users that they are no longer a part of these discussions/topics and should contact admin to either invite you to the stream itself or shift the topic to a new stream which you can subscribe to, if your presence is important in the discussion?

Maybe we can create an additional bot user with some name like `Zulip importer bot` in the imported organization and send some direct messages to the user from this bot account that they are no longer subscribed to so and so discussions?

hackerkid commented:
>But if the channel is private, wouldn't that lead them to access the messages they are not supposed to see? 

hmm. How does the feature work exactly in Rocket.Chat? Can users be added to only certain discussions instead of adding to the entire stream or is it more like the user choosing to only follow certains discussions in a channel which they are already added? 

if it's the former, then we can skip the subscriptions for private streams and keep the public ones. Yeah. Something like the bot will also work. But I imagine we need to modify our import tool to support this particular use case. 

I can't really seem to find anything useful in their docs. They also seems to also have "threads" along with discussions. How does that fit in here?

garg3133 commented:
> Can users be added to only certain discussions instead of adding to the entire stream or is it more like the user choosing to only follow certains discussions in a channel which they are already added? Who decides what discussions the user have access to in private streams?

I tried playing with the discussions a bit and found out that both parts of the first statement above are true. Initially, the person creating a discussion decides who is going to be a part of that discussion (invitees may or may not be subscribed to the parent channels). But after a discussion is created, anyone who has access to that channel can join the discussion voluntarily. So, for public channels, anyone can join a discussion (even without joining the parent channel) and for private channels, only the people who are subscribed to that channel can join the discussion themselves (others would need to be invited to join the discussion).

> then we can skip the subscriptions for private streams and keep the public ones

Yeah, this makes sense.

> They also seems to also have "threads" along with discussions. How does that fit in here?

When they created discussions, they were initially called Threads. But I guess when the users didn't like it much, they changed its name to Discussions and just copied Slack's implementation of Threads. More details [here](https://forums.rocket.chat/t/doesnt-make-sense-to-have-both-thread-and-discussion/3711/11).

For threads, currently we just mix them with the normal messages according to timestamps, but we'll be making them as separate topics in a follow-up of this PR.

timabbott commented:
I think given the complexity of this PR at this point, it makes sense to merge what we have and make further improvements in follow-ups.  I've done so after a quick pass on the documentation.  My top priorities are the following:
* The documentation should spell out the exact commands required to get the Rocket.Chat database dump into the right format, including a `tar` command to generate something that they can send to the Zulip Cloud team (and can also unpack in the self-hosted instructions).
* From experience talking to users of past import tools, custom emoji are probably the most important missing feature, followed by attachments.

I'm going to delay advertising that this feature exists for anything other than early testing/QA while we do near-term improvements.

Huge thanks for adding this!
